book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,218 @@
+all frozen ancestors `>= R`. Furthermore, `BankForks` prunes ancestors in its
+set of `banks` that are `< R'` after setting a new root `R'`, so this invariant
+is always true.
+
+**Lemma 2:** Given a root bank `R` of `BankForks` and some `L_i` in locktower,
+if `L_i >= S`, then `L_i >= R`.
+
+**Proof:** When we boot from a snapshot, we set `R == S`, and in this case
+because we are assuming `L_i > S`, then we know `L_i >= R` at bootup. Then when
+we set a new root in BankForks `R'` after booting, we guarantee this only
+happens if the locktower root is also set to `R'`. By the construction of
+locktower, all votes in the tower must be greater than the locktower root, so
+`L_i >= R'`. 
+
+For `Case 2` we assumed `L_i` >= `S`, so from Lemma 2 we know `L_i` >= `R`.
+Then from Lemma 1 we know its sufficient to check
+`B_d.ancestors().contains(L_i)`.
+
+
+## Boot Procedure:
+
+1) The validator gets a snapshot to boot from. This can come from any source,
+as it will be verified before being vote upon. Call the root of this snapshot
+`S`.
+
+2) We reject any snapshots `S` where `S` is less than the root of `L` and `S`
+is not in the list of roots in blocktree because this means `S` is for a
+different fork than the one this validiator last rooted. Thus it's critical for
+consistency in `replay_stage` that the order of events when setting a new root
+is:
+
+    1) Write the root to locktower 2) Write the root to blocktree 3) Generate
+    snapshot for that root
+
+3) On startup, the validator checks that the current root in locktower exists
+in blocktree. If not (there was a crash between 2-1 and 2-2), then rewrite the
+root to blocktree.
+
+4) On startup, the validator boots from the snapshot `S`, then replays all
+descendant blocks of `S` that exist in this validator's ledger, building banks
+which are then stored in an output `BankForks`. This is done in
+`blocktree_processor.rs`. The root of this `BankForks` is set to `S`.
+
+5) On startup, the validator calculates the first slot `S_n` from which it is
+not locked out for voting. Every validator persists its locktower state and
+must consult this state in order to boot safely and resume from a snapshot
+without being slashed. From this locktower state and the ancestry information
+embedded in the snapshot, a validiator can derive which banks, are "safe" (See
+the `Determining Vote Safety From a Snapshot and Locktower` section for more
+detals) to vote for.
+
+6) Periodically send canary transactions to the cluster using the validator's
+local recent blockhash.
+
+7) Wait for the following criteria. While waiting, set a new root every time
+2/3 of the cluster's stake roots a bank. This allows the validator to prune its
+state and also calculate leader schedules as it moves across epochs.
+
+   1) Observe a canary transaction in a bank that some threshold of trusted
+   validator stake has voted on. Call this trusted bank `T`.
```

Comment:
Probably don't need this `T` as its not referenced anywhere

Reply:
It's referenced on line 37.